src/loader.cpp

```diff
@@ -0,0 +1,55 @@
+#include "loader.hpp"
+
+InstructionLoader::InstructionLoader(const std::string &trace_filename) {
+    LoadTraceFile(trace_filename);
+}
+
+InstructionLoader::~InstructionLoader() {
+    if (in_file == nullptr) {
+        in_file->close();
+    }
+}
+
+void InstructionLoader::LoadTraceFile(const std::string &filename) {
+    in_file = std::make_unique<std::ifstream>(filename.c_str(), std::ios::in);
+    if (in_file->fail()) {
+        std::cerr << "Open trace file error" << std::endl;
+        exit(-1);
+    }
+}
+
+inst_t InstructionLoader::GetNextInst() {
+    const int LENGTH_OF_INST_LINE = 13;
+    char trace_line[13];
+    if (IfAvailable())
+        in_file->getline(trace_line, LENGTH_OF_INST_LINE);
+
+    return _ParseLineToInst(trace_line);
+}
+
+bool InstructionLoader::IfAvailable() { return !(in_file->eof()); }
+
+inst_t InstructionLoader::_ParseLineToInst(const char *line) {
+    int INST_ADDR_BASE = 16;
+    addr_raw_t addr = (addr_raw_t)strtoul(line + 2, nullptr, INST_ADDR_BASE);
+
+    switch (line[0]) {
+    case 'l':
+        // TODO: add statistics
+        return inst_t(I_LOAD, addr);
+        break;
+    case 's':
+        // TODO: add statistics
+        return inst_t(I_STORE, addr);
+        break;
+    case '\0':
+        // TODO: add statistics
+        return inst_t(I_NONE, addr);
+        break;
+    default:
+        return inst_t(I_NONE, addr);
+        std::cerr << "Undefined instruction type." << std::endl;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Statements following return, break, continue, goto or throw will never be executed.](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103)

Reply:
@jackyhuang85  in default branch, error message should be printed before return